internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -192,42 +215,186 @@ func (r *NumaflowControllerRolloutReconciler) reconcile(
 	ctx context.Context,
 	controllerRollout *apiv1.NumaflowControllerRollout,
 	namespace string,
-) error {
+) (ctrl.Result, error) {
 	numaLogger := logger.FromContext(ctx)
 
 	if !controllerRollout.DeletionTimestamp.IsZero() {
 		numaLogger.Info("Deleting NumaflowControllerRollout")
 		if controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
+			GetPauseModule().deleteControllerPauseRequest(namespace)
 			controllerutil.RemoveFinalizer(controllerRollout, finalizerName)
 		}
 		// generate the metrics for the numaflow controller deletion based on a numaflow version.
 		r.customMetrics.DecNumaflowControllerMetrics(controllerRollout.Name, controllerRollout.Namespace, controllerRollout.Spec.Controller.Version)
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	// add Finalizer so we can ensure that we take appropriate action when CRD is deleted
 	if !controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
 		controllerutil.AddFinalizer(controllerRollout, finalizerName)
 	}
 
-	// apply controller
+	_, pauseRequestExists := GetPauseModule().getControllerPauseRequest(namespace)
+	if !pauseRequestExists {
+		// this is just creating an entry in the map if it doesn't already exist
+		GetPauseModule().newControllerPauseRequest(namespace)
+	}
+
+	deployment, deploymentExists, err := r.getNumaflowControllerDeployment(ctx, controllerRollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if deploymentExists {
```

Comment:
Can they be a race condition, that `deploymentExists` returns false during checking, but it become true when `sync` happens and we have a version upgrade?

Reply:
oh, are you referring to the case in which somebody manually deleted the Numaflow Controller Deployment?

Reply:
that could be one case.

Reply:
@xdevxy  Thinking about this one more. I think `deploymentExists` can happen in 3 cases:
- the `NumaflowControllerRollout` has just been created: we won't try to pause and that seems okay (theoretically, there could be Pipelines running and `NumaflowControllerRollout` has been removed and recreated, but I don't think we can guarantee anything in that case, unless we tried to figure out which version of numa container was running on all of the Pipelines)
- the `NumaflowControllerRollout` was created in the previous reconciliation and the Informer cache is old: if by some rare chance, the `NumaflowControllerRollout` also just updated its version at the very same time, we won't try to pause and that seems okay as the NumaflowController only just got rolled out anyway
- somebody manually deleted the Numaflow Controller `Deployment`: if by some rare chance, the  `NumaflowControllerRollout` also just updated its version at the very same time, we won't try to pause - this case wouldn't be great but it's so rare, it seems kind of low priority. If we did try to handle it, I would just imagine that we could store the previous version in the NumaflowControllerRollout's Status and compare to that. Want me to maybe just add an issue for this, or how do you want me to move forward?

Reply:
yeah I think these are rare cases so let's consider it low priority and having issue to track this for now.